### ISSUE TITLE
'apply' for static DHCP mapping changes

### ISSUE THREAD

**Issue by codesmaker** [jaredhendrickson13/pfsense-api#316]
Hi,

Thank you for this awesome project!

I'm running pfSense 2.6.0 and I successful run an API call to add a static DHCP mapping. The problem is that it's not showing in the WebUI. I search for an 'apply' call for DHCP but found nothing. I tried to restart the DHCP server but I still couldn't see it.

Any idea?

**Comment by jaredhendrickson13** [comment]
Hey!

There currently isn't an apply endpoint or field for the DHCP server/static mappings, rather the configuration is applied immediately upon successful API call for these endpoints. That being said, the static mapping endpoint appears to be working as expected for me with API v1.5.1 on pfSense 2.6.0. 

Based on the behavior you describe, it sounds like the user you are using to authenticate may have the `user-config-readonly` (User: Config: Deny Config Write as it shows in the webConfigurator) privilege assigned. This will make the API call appear successful, but nothing actually gets written to the configuration; therefor nothing would show in the webConfigurator. If that's the case, removing that privilege should fix the issue.

If that privilege is not currently assigned, can you provide examples of your API requests and responses along with what shows in the webConfigurator afterwards?

Thanks!

**Comment by codesmaker** [comment]
Hey @jaredhendrickson13!

Thanks a lot for your answer and help.

Actually, I'm using an account with administrator-level access. Here are all the privileges that are currently assigned to the user:
![image](https://user-images.githubusercontent.com/6999196/219132273-5026b25a-1b11-435d-af3e-20c8b9ccd1cc.png)

Here is the API request:
 curl -XPOST 'https://pfsense.example.com/api/v1/services/dhcpd/static_mapping' -H 'accept: application/json' -H 'Content-Type: application/json' -u admin:mypassword -d @new-ip.txt -k

And here is the response:
{"status":"ok","code":200,"return":0,"message":"Success","data":{"mac":"52:23:00:46:FE:72","ipaddr":"192.168.0.19","descr":"Created by a script","hostname":"MY-TEST-HOST"}}

When I check in the UI, the static mapping I just added with the IP '192.168.0.19' is not showing and the 'apply' button is not there. In order for it to appear, I modify any static mapping that I have and then save. The 'apply' button appears and when I click it and the UI reloads, the mapping is there.

I'm also running the API v1.5.1 on pfSense v2.6.0.

**Comment by codesmaker** [comment]
Can the 'apply' be triggered programmatically?

**Comment by jaredhendrickson13** [comment]
The 'apply' action is performed anytime a successful POST, PUT or DELETE request is made to /api/v1/services/dhcpd/static_mapping, you could try making a PUT request to update the static mapping to re-apply it but I'm not sure that's the problem here. The events that occur during the API call are:

1. The mapping entry is validated. If a validation error occurs, an API error will be returned.
2. The validated mapping entry is written to the pfSense config.
3. The DHCP server configuration is applied using the values in the pfSense config.

Any chance you could provide samples of the pfSense XML config contents before you make the request, after you make the request, and then again after you apply from the webConfigurator? The entry not showing in the webConfigurator after making the original request implies the mapping is not actually being written to the config file, but it showing up after applying implies that it actually is there but for some reason is not being seen by the webConfigurator.

**Comment by jaredhendrickson13** [comment]
Thanks for providing that!

Odd enough, this shows that the API is writing the static mapping to config as expected. The webConfigurator just reads what is present in the XML configuration file, even if the config hasn't been applied to the DHCP service, so this just shows that it should be working. I'm assuming you are refreshing the 'Services > DHCP Server' webConfigurator page after you make the API call?

I'm still having trouble replicating the problem. Here is what I'm doing to try to replicate/debug the issue on a fresh install of pfSense 2.6.0:

1) On a fresh install of 2.6.0, I install the latest API package (v1.5.1) using the following command:
```
pkg add https://github.com/jaredhendrickson13/pfsense-api/releases/latest/download/pfSense-2.6-pkg-API.txz && /etc/rc.restart_webgui
```

2) I ensure the package is running v1.5.1 with this command:
```
pfsense-api version
```

Which shows:
```
pfSense-pkg-API-1.5_1
Name           : pfSense-pkg-API
Version        : 1.5_1
Installed on   : Thu Feb 16 15:50:56 2023 UTC
Origin         : sysutils/pfSense-pkg-API
Architecture   : FreeBSD:12:amd64
Prefix         : /usr/local
Categories     : sysutils
Licenses       : APACHE20
Maintainer     : github@jaredhendrickson.com
WWW            : https://github.com/jaredhendrickson13/pfsense-api
Comment        : pfSense API package
Annotations    :
	FreeBSD_version: 1200086
Flat size      : 11.4MiB
Description    :
The missing REST API package for pfSense.
WWW: https://github.com/jaredhendrickson13/pfsense-api
```

3) I run the following curl command to add the static entry:
```
curl -k -u admin:pfsense -d '{"mac": "52:23:00:46:FE:76", "ipaddr": "192.168.1.3", "descr": "Created by a script", "hostname": "MY-TEST-HOST", "interface": "lan"}' -X POST https://127.0.0.1/api/v1/services/dhcpd/static_mapping
```

Which responds with:
```json
{"status":"ok","code":200,"return":0,"message":"Success","data":{"mac":"52:23:00:46:FE:76","ipaddr":"192.168.1.3","descr":"Created by a script","hostname":"MY-TEST-HOST"}}
```

4) I login to the webConfigurator and check the 'Services > DHCP Server' page on the LAN interface tab and see the following under the 'Static mappings' section:
![Screen Shot 2023-02-16 at 5 10 31 PM](https://user-images.githubusercontent.com/50180862/219516418-40a88706-caed-40f5-bd4f-816c4bfd1f4e.png)

5) I then check the dhcpd.conf file to ensure the static mapping was applied to the DHCP service by the API:
```
cat /var/dhcpd/etc/dhcpd.conf
```

which shows the mapping has been applied:
```
option domain-name "home.arpa";
option ldap-server code 95 = text;
option domain-search-list code 119 = text;
option arch code 93 = unsigned integer 16; # RFC4578

default-lease-time 7200;
max-lease-time 86400;
log-facility local7;
one-lease-per-client true;
deny duplicates;
update-conflict-detection false;
authoritative;
class "s_lan" {
	match pick-first-value (option dhcp-client-identifier, hardware);
}
subnet 192.168.1.0 netmask 255.255.255.0 {
	pool {

		range 192.168.1.100 192.168.1.199;
	}

	option routers 192.168.1.1;
	option domain-name-servers 192.168.1.1;
	ping-check true;

}
host s_lan_0 {
	hardware ethernet 52:23:00:46:FE:76;
	fixed-address 192.168.1.3;
	option host-name "MY-TEST-HOST";

}
subclass "s_lan" 1:52:23:00:46:FE:76;
```

Is there a step I'm missing or maybe something unique to your pfSense instance that isn't present on a fresh install of pfSense?

**Comment by codesmaker** [comment]
You're welcome! I finally had some time to further testing the issue. I installed a fresh pfSense and tried to added a DHCP static mapping and it worked as expected. So I thought about reboot the pfSense in question to make sure that there was nothing funny going on and be hold, it worked :). So I guess it was just a temporary thing and the reboot fixed it.

Thanks a lot again for your time and support! Please keep up the good work!
